Update spec for compatible implementation definition. (#215)
eclipse/microprofile: spec/src/main/asciidoc/required-apis.asciidoc

 - <<mp-opentracing, MicroProfile OpenTracing 2.0>>
 - <<mp-rest-client, MicroProfile Rest Client 2.0>>

A MicroProfile 4.x compatible implementation must pass all the TCKs provided by the MicroProfile individual

> specifications. The Jakarta specifications TCKs are not required.

Reviewer: I would say: The Jakarta specifications TCKs are not required. => The Jakarta specifications TCKs are optional.

Reviewer 2: I'd remove the "4.x" verbiage. First, it does not apply to 4.0. Second, this spec applies to 4.1 by virtue of being in the 4.1 platform specification.  Recommended verbiage - "MicroProfile compatible implementations must pass ...".

These APIs and versions guarantee application developers a minimum set of APIs and are not intended to be restrictive.

> MicroProfile applications that make use of APIs or versions not mentioned here are not considered portable.

Reviewer 3: What does portable mean in this context?

Author: My interpretation of this (and this text was already in the spec, I just moved it around), is that if an application uses any API (not an implementation, for instance Bean Validation), that is not on the list then the application may not be able to run in all MP available runtimes.

Reviewer: I would delete the whole sentence "MicroProfile applications that make use of APIs or versions not mentioned here are not considered portable." as the previous sentence has hinted. Besides, they might be portable if they move among the runtimes that support the same.

Reviewer 3: Sure; and I don't want to hijack the pull request.  But for "compatible implementation" purposes I do think _somewhere_ it is important to say something rather in depth about portability. For example, there is no world that I'm aware of in which a Helidon application will run on or in OpenLiberty, and no world in which a Payara application will run in Quarkus, and so on etc. If that's OK, as the intent behind this compatibility effort seems to be, then users are probably owed a paragraph or two explaining the kind of portability MicroProfile is talking about.

Author: @ljnelson do you think we should just remove this part?

Reviewer 2: I find this section somewhat confusing, tbh. Regarding the platform spec, I would say a better approach is "The behavior and APIs encompassed by this specification are considered consistent across compatible implementations." OK, not the best wording, but using the word "portable" doesn't really fit. Interestingly, we use the phrase "Platform Spec" when MicroProfile isn't really a standalone platform because of its relatively narrow focus. If we really want to use the word "portable", the context has to be narrowed to what it "does do" (and not "what it does not do").

I'd be good with removing this section altogether.

Reviewer 4: Let me give a little bit of history of this section.  We wanted to document the minimum version of each of our dependencies.  We didn't want to lock applications into these specific versions.  If developers used these defined minimum versions, then the resulting MP applications would be portable between implementations.  But, if a newer version of one of these dependencies were used, all bets were off.  Depending on the changes introduced by this newer version, the application may or may not be portable.

For example, the defined dependency for CDI is CDI 2.0.  If CDI 2.0 is used by an application, then the app should be portable to other MP implementations.  But, if CDI 2.1 is used by an application, the app may or may not be portable.

That was the intent of this section.  We wanted to defined the minimum version for all of our dependencies, but not lock developers into these specific versions.  The caveat to this approach is that we don't know what the newer versions of the dependencies would introduce, so we can't claim portability with the unknown.

Reviewer: I think we are better off not to assert either portable or non-portable. Removing the added sentence talking about `portable` sounds good to me.

Reviewer 3: I tend to agree with @Emily-Jiang here: since MicroProfile has no notion of portability—you can't write a MicroProfile application and "run" or "deploy" it on any two MicroProfile implementations, unmodified—it probably shouldn't be mentioned here (or elsewhere).  (You can of course "port" your knowledge of and familiarity with the various APIs from one runtime to another as you re-assemble your application dedicated to that runtime, but that's different.)  To the extent that it is collectively believed that MicroProfile _should_ have some kind of notion of portability, that should be addressed as a separate (probably enormous 😄 ) effort.  OK, done hijacking this thread.

Author: I'm ok to remove this.

[[javase]]
=== Java SE

Libraries in use by the MicroProfile leverage new features provided in Java SE 8, therefore Java SE 7 and before are unsupported.

> MicroProfile 4.x implementations are free to support any Java SE version.

Reviewer: ```suggestion
MicroProfile 4.x implementations are free to provide support for Java SE 8 and beyond.
```

Reviewer 4: Personally, I don't understand why the original statement has to change at all.  It says exactly what we were saying on the call.  Go ahead and use any version of Java SE beyond 8 (inclusive), as long as you Java SE 8 compatibility is maintained.

Author: I interpret the part `however Java SE 8 compatibility must be maintained.` as the implementation must support a Java SE 8 runtime. Is this a wrong interpretation?

Also, my understanding from the discussion and consensus in the call was to relax this requirement, and allow implementations to support whatever version they wanted, but without the restriction to always have to support 8.

 - <<mp-opentracing, MicroProfile OpenTracing 2.0>>
 - <<mp-rest-client, MicroProfile Rest Client 2.0>>

> A MicroProfile compatible implementation must pass *all* TCKs provided by each MicroProfile specification. Passing the

Reviewer 5: Maybe also mention MP TCKs that are optional? If we have something like this in some of the platform specs. Not sure, I would need to check.

Author: Ah, you mean the optional tests for EJB integration in JWT for instance?

Reviewer 5: @radcortez exactly. Also, there might be more optional TCKs in the future.

Jakarta specifications TCKs is optional.

> These APIs and versions guarantee application developers a minimum set of APIs and are not intended to be restrictive.

Reviewer: ```suggestion
These APIs and versions provide application developers a minimum set of APIs, but they are not intended to be restrictive.
```

Reviewer 5: I would keep guarantee as this meant as in guarantee portability for these APIs.

[[javase]]
=== Java SE

Libraries in use by the MicroProfile leverage new features provided in Java SE 8, therefore Java SE 7 and before are
unsupported.

MicroProfile APIs and specifications use Java SE 8 for both source and target compatibility. MicroProfile

> implementations are free to provide support for runtimes starting at Java SE 8 and higher.

Reviewer: ```suggestion
implementations are free to provide support for runtimes starting at Java SE 8 or higher.
```

 - <<mp-opentracing, MicroProfile OpenTracing 2.0>>
 - <<mp-rest-client, MicroProfile Rest Client 2.0>>

> A MicroProfile compatible implementation must pass *all* the TCKs provided by the MicroProfile individual

Reviewer: *all* the TCKs => *all* TCKs

specifications. The Jakarta specifications TCKs are optional.

These APIs and versions guarantee application developers a minimum set of APIs and are not intended to be restrictive.

[[javase]]
=== Java SE

Libraries in use by the MicroProfile leverage new features provided in Java SE 8, therefore Java SE 7 and before are unsupported.

> MicroProfile implementations are free to provide support for any Java SE 8 and beyond.

Reviewer 6: beyond > higher

[[jakarta-cdi]]
=== Jakarta Contexts and Dependency Injection (CDI) 2.0

Jakarta CDI provides the base for a growing number of APIs included in MicroProfile.
Use of implementations beyond CDI 2.0 in MicroProfile is allowed, however, not required.

 - https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html
 - http://cdi-spec.org/

==== Optional Requirements

> The following CDI features are optional:

Reviewer: Since we don't have the complete list on what are optional, I suggest instead of highlighting the list, we should just say "Some of CDI features are not required, such as EJB integration, Portable Extension, Expression langages etc.". Please refer to the individual implmentation on their support statement.

 - @ConversationScoped
 - Expression Language
 - EJB Integration
 - Portable Extensions

[[jakarta-jaxrs]]
=== Jakarta RESTful Web Services (JAX-RS) 2.1

Jakarta JAX-RS provides both standard client and server APIs for RESTful communication by MicroProfile applications.
Use of implementations beyond JAX-RS 2.1 in MicroProfile is allowed, however, not required.

 - https://jakarta.ee/specifications/restful-ws/2.1/restful-ws-spec-2.1.html
 - https://github.com/eclipse-ee4j/jaxrs-api

==== Optional Requirements

> The following JAX-RS features are optional:

Reviewer: Since we don't have the complete list on what are optional, I suggest instead of highlighting the list, we should just say "Some of JAX-RS features are not required, such as Servlet integration, Managed Beans Integration, EJB Integration, etc.". Please refer to the individual implmentation on their support statement.

 - https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html
 - http://cdi-spec.org/

==== Optional Requirements

> MicroProfile 4.x implementations may opt to not implement the following CDI features:

Reviewer: ```suggestion
The following CDI features have been made optional in the MicroProfile 4.x implementations.
```

Reviewer 4: We can't declare portions of CDI as being optional.  The Jakarta EE CDI team can declare parts of their spec as optional, but MicroProfile can't do that.  We can state the support of CDI in general is optional (per the earlier TCK test requirements).  But, we can't declare portions of the CDI API or Annotations as being optional.

Reviewer 6: Is the scenario that we declare these features optional just in the context of MP possible? Meaning that we "override" this in the TCKs that are being used for MP, not touching the underlying Jakarta EE base?

Reviewer: I don't think we can claim CDI as being optional though. CDI is required. EJB, EL part from CDI are optional as we removed that dependency.

Reviewer: I think we need to reword: some part of CDI specs are not required in MicroProfile.

Author: My understanding from the discussion call was that we were going to list part of the CDI spec that were not required by MP. As stated, it is impossible for a MP only implementation to pass the CDI TCK, because it includes things like EL Support and EJB integration that we explicitly exclude in the specification pom.

In my opinion we are already making these things optional (from the pom), we are just specifying them here. I'm ok to go over this discussion again, but in my opinion the specification pom and the specification wording need to align.

Reviewer: By the way, I disagree with the above statement @radcortez . Even though some features such EL, EJB were excluded, runtimes are absolutely fine to add them in their integration. MP should not force MP impls not to integrate EL, EJB, etc. I think we should only assert on these integrations are not required.

Author: To clarify, all my considerations are about a MP runtime without additional Jakarta dependencies.

If we say these are optional (or not have to be included), is not forcing implementations to not include these features.

 - https://jakarta.ee/specifications/restful-ws/2.1/restful-ws-spec-2.1.html
 - https://github.com/eclipse-ee4j/jaxrs-api

==== Optional Requirements

> MicroProfile 4.x implementations may opt to not implement the following JAX-RS features:

Reviewer: ```suggestion
The following JAX-RS features have been made optional in the MicroProfile 4.x implementations.
```

Reviewer 4: Same comment here...  MicroProfile can not declare portions of the JAX-RS Spec as being optional.  Only the JAX-RS team can declare those type of things.

Reviewer: It might be a bit sticky on JAX-RS when moving forward. In the current Jakarta EE 10, there is no plan to split JAX-RS to address the requirement for MP. Even though Jakarta Core Profile 1.0 is released, MP might still not be able to consume it. @starksm64 do we need a JAX-RS lite to be included in Jakarta Core Profile?
